src/rules/core/statistic.rs

```diff
@@ -0,0 +1,223 @@
+#[derive(Debug, Clone)]
+pub struct Statistic {
+    id: StatisticId,
+    value: StatisticValue,
+}
+
+impl Statistic {
+    fn new(id: StatisticId, value: StatisticValue) -> Self {
+        Self { id, value }
+    }
+
+    accessor! {race, Race, RaceId}
+
+    accessor! {class, Class, ClassId}
+
+    accessor! {level, Level, Level}
+
+    accessor! {hit_points, HitPoints, HitPoints}
+
+    accessor! {ability, Ability, AbilityScore}
+
+    accessor! {skill, Skill, Proficiency}
+
+    accessor! {proficiency_bonus, ProficiencyBonus, ProficiencyBonus}
+
+    /// Applies a change on this statistic. The change will have an effect only if its
+    /// type matches the statistic's one.
+    #[allow(dead_code)] // TODO remove when used
+    pub(crate) fn apply_change(&mut self, change: &StatisticChange) {
+        self.value.apply_change(change);
+    }
+}
+
+impl From<StatisticInitializer> for Statistic {
+    fn from(item: StatisticInitializer) -> Self {
+        use StatisticInitializer::*;
+        match item {
+            Race(value) => Statistic::new(StatisticId::Race, StatisticValue::Race(value)),
+            Class(value) => Statistic::new(StatisticId::Class, StatisticValue::Class(value)),
+            Level(value) => Statistic::new(StatisticId::Level, StatisticValue::Level(value)),
+            HitPoints(value) => {
+                Statistic::new(StatisticId::HitPoints, StatisticValue::HitPoints(value))
+            }
+            Ability(id, value) => {
+                Statistic::new(StatisticId::Ability(id), StatisticValue::Ability(value))
+            }
+            Skill(id, value) => {
+                Statistic::new(StatisticId::Skill(id), StatisticValue::Skill(value))
+            }
+            ProficiencyBonus(value) => Statistic::new(
+                StatisticId::ProficiencyBonus,
+                StatisticValue::ProficiencyBonus(value),
+            ),
+        }
+    }
+}
+
+impl weasel::Id for Statistic {
+    type Id = StatisticId;
+
+    fn id(&self) -> &Self::Id {
+        &self.id
+    }
+}
+
+/// Uniquely identifies a `Statistic`.
+#[derive(Debug, Clone, Copy, PartialEq, Hash, Eq, Serialize, Deserialize)]
+#[allow(missing_docs)]
+pub enum StatisticId {
+    Race,
+    Class,
+    Level,
+    HitPoints,
+    Ability(AbilityId),
+    Skill(SkillId),
+    ProficiencyBonus,
+}
+
+/// Encapsulates the actual value of a statistic.
+#[derive(Debug, Clone)]
+enum StatisticValue {
+    Race(RaceId),
+    Class(ClassId),
+    Level(Level),
+    HitPoints(HitPoints),
+    Ability(AbilityScore),
+    Skill(Proficiency),
+    ProficiencyBonus(ProficiencyBonus),
+}
+
+impl StatisticValue {
+    fn apply_change(&mut self, change: &StatisticChange) {
+        let mut successful = false;
+        use StatisticValue::*;
+        match change {
+            StatisticChange::HitPoints(change) => {
+                if let HitPoints(value) = self {
+                    if *change > 0 {
+                        value.add(change.abs() as u16);
```

Comment:
the r-value here could be moved to a variable to make it more DRY

Reply:
done